configcat/php7-sdk
# Config v6

Implements config v6 support.

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.getMatchedTargetingRule`/`getMatchedPercentageOption` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValue`

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.getMatchedTargetingRule`/`getMatchedPercentageOption` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)
* [x] Comparison attribute and comparison value trimming tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute conversion to canonical string representation tests (https://trello.com/c/CZgGrrcb)

Related improvements/fixes:
* [x] Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)
* [ ] ~We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)~ (PHP SDK doesn't support polling modes)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)

Additional fixes:
* Fix inconsistent logging of error 1000 in the case of flag overrides.

Additional improvements:
* Improve consistency of config JSON deserialization error reporting.
* Add exception message and stack trace to the log format of the default logger.
* When evaluation fails and there is a related `Throwable` object, also expose it `EvaluationDetails` via the new `getErrorException` property.

Breaking changes:

- Rename the `getError` property to `getErrorMessage` in `EvaluationDetails`.
- Rename the `getError` property to `getErrorMessage` in `RefreshResult`.
- Add an extra parameter of type `?Throwable` to the `onError` hook callbacks.
- Rename the `getMatchedEvaluationRule` property to `getMatchedTargetingRule` and the `getMatchedEvaluationPercentageRule` property to `getMatchedPercentageOption` in `EvaluationDetails`.
- Move config JSON attribute descriptor classes from the `ConfigCat\Attributes` to the `ConfigCat\ConfigJson` namespace and update them to support [config JSON schema v6](https://github.com/configcat/config-json/tree/main/V6).
- Don't include `event_id` in the log message but leave it up to the actual logger implementation to choose the log message format. (Affects consumers using custom logging only.)

### Requirement checklist (only if applicable)
